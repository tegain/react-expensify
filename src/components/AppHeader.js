import React from 'react'
import { NavLink } from 'react-router-dom'

const AppHeader = () => (
	<header>
		<h1>Expensify</h1>
		<ul>
			<li><NavLink to="/" activeClassName="is-active" exact={true}>Dashboard</NavLink></li>
			<li><NavLink to="/create" activeClassName="is-active">Add new expense</NavLink></li>
			<li><NavLink to="/help" activeClassName="is-active">Help</NavLink></li>
		</ul>
	</header>
)

export default AppHeader