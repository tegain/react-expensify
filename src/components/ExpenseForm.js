import React from 'react'
import moment from 'moment'
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates'

export default class ExpenseForm extends React.Component {

	constructor (props) {
		super (props)

		this.state = {
			description: props.expense ? props.expense.description : '',
			note: props.expense ? props.expense.note : '',
			amount: props.expense ? (props.expense.amount / 100).toString() :'',
			createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
			calendarFocused: false,
			error: undefined
		}
	}

	onDescriptionChange = (e) =>Â {
		const description = e.target.value
		this.setState(() => ({ description }))
	}

	onNoteChange = (e) => {
		const note = e.target.value
		this.setState(() => ({ note }))
	}

	onAmountChange = (e) => {
		const amount = e.target.value

		/**
		 * Starts (^) with digit no matter how long (but at least one, to infinity): \d{1,}
		 * Then takes optional group ()?, allowing 0 or 1 point, followed by 0 to 2 digits: (\.\d{0,2})?
		 * then finishes the expression: $
		 * @type {RegExp}
		 */
		const regex = /^\d{1,}(\.\d{0,2})?$/
		if (!amount || amount.match(regex)) { // '!amount' allows empty value, so the user can delete the amount
			this.setState(() => ({ amount }))
		}
	}

	onDateChange = (createdAt) => {
		if (createdAt) {
			this.setState(() => ({ createdAt }))
		}
	}

	onFocusChange = ({ focused }) => {
		console.log({focused})
		this.setState(() => ({ calendarFocused: focused }))
	}

	onSubmit = (e) => {
		e.preventDefault()

		if (!this.state.description || !this.state.amount) {
			// Set error
			this.setState(() => ({ error: 'Please provide description and amount' }))
		} else {
			// Clear error
			this.setState(() => ({ error: undefined }))

			/**
			 * Dispatch to prop method, so the Redux action actually happens on the AddExpensePage,
			 * and not on this form (which is supposed to be reusable for AddExpense and EditExpense)
			 */
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount) * 100,
				createdAt: this.state.createdAt.valueOf(), // Get timestamps from moment.js method
				note: this.state.note
			})
		}
	}

	render () {
		return (
			<div className="input-group">
				{this.state.error && <p>{this.state.error}</p>}

				<form onSubmit={this.onSubmit} className="expenses-form">
					<div className="input-group__item">
						<input
							type="text"
							className="text-input"
							placeholder="Description"
							autoFocus
							value={this.state.description}
							onChange={this.onDescriptionChange}
						/>
					</div>

					<div className="input-group__item">
						<input
							type="text"
							className="text-input"
							placeholder="Amount"
							value={this.state.amount}
							onChange={this.onAmountChange}
						/>
					</div>

					<div className="input-group__item">
						<SingleDatePicker
							date={this.state.createdAt}
							onDateChange={this.onDateChange}
							focused={this.state.calendarFocused}
							onFocusChange={this.onFocusChange}
							numberOfMonths={1}
							isOutsideRange={(day) => false} // Enables to select dates in the past
						/>
					</div>

					<div>
						<textarea
							placeholder="Add a note for your expense (optional)"
							value={this.state.note}
							onChange={this.onNoteChange}
						>
						</textarea>
					</div>

					<div>
						<button
							type="submit"
							className="button button--blue"
						>
							Save Expense
						</button>
					</div>
				</form>
			</div>
		)
	}

}