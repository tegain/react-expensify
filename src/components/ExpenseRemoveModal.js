import React from 'react'
import { connect } from 'react-redux'
import Modal from 'react-modal'
import {startRemoveExpense} from "../actions/expensesActions";

export class ExpenseRemoveModal extends React.Component {
	onRemoveExpense = () => {
		this.props.startRemoveExpense({ id: this.props.removeExpenseId })
		this.props.history.push('/')
	}

	render () {
		return (
			<Modal
				appElement={document.getElementById('root')}
				isOpen={!!this.props.removeExpenseId}
				onRequestClose={this.props.onClearRemoveExpense}
				contentLabel="Remove expense"
				closeTimeoutMS={200}
				className="modal"
			>
				<h3 className="ReactModal__title">Remove an expense</h3>
				<div className="ReactModal__inner">
					<p>Remove expense <strong>{this.props.removeExpenseDesc}</strong> ?</p>
					<div className="ReactModal__actions">
						<button
							className="button button--blue"
							onClick={this.onRemoveExpense}
						>
							Remove
						</button>

						<button
							className="button button--transparent"
							onClick={this.props.onClearRemoveExpense}
						>
							Cancel
						</button>
					</div>
				</div>
			</Modal>
		)
	}
}

const mapDispatchToProps = (dispatch) => ({
	startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
})

export default connect(undefined, mapDispatchToProps)(ExpenseRemoveModal)