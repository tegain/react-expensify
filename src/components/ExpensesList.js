import React from 'react'
import { connect } from 'react-redux'
import ExpensesListItem from './ExpensesListItem'
import selectExpenses from '../selectors/expensesSelectors'

// Takes props from the Connected Component
export const ExpensesList = (props) => (
	<div className="expenses">
		<div className="expenses-header">
			<span className="expenses-header__expense">Expense</span>
			<span className="expenses-header__amount">Amount</span>
		</div>

		<div className="expenses-list">
			{
				props.expenses.length === 0 ? (
					<p className="expenses-list__empty">No expenses</p>
				) : (
					props.expenses.map((expense) => (
						<ExpensesListItem key={expense.id} {...expense} /> // Deconstruct expenses properties
					))
				)
			}
		</div>
	</div>
)

/**
 * This is a function that wraps your component in another component,
 * which subscribes to changes in the Redux store and renders itself and
 * consequently its descendants whenever an update occurs.
 *
 * connect(what_we_want_to_connect)(which_component_we_want_to_connect)
 */
// const ConnectedExpensesList = connect((state) => {
// 	return {
// 		expenses: state.expenses // We only want to connect state.expenses (not all the state)
// 	}
// })(ExpensesList)
//
// export default ConnectedExpensesList

// REFACTOR -->

const mapStateToProps = (state) => {
	return {
		expenses: selectExpenses(state.expenses, state.filters) // We only want to connect the filtered state.expenses (not all the state)
	}
}

export default connect(mapStateToProps)(ExpensesList)

