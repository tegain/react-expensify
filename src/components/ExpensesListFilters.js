import React from 'react'
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates'
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filtersActions'

/**
 * Connected component has access to dispatch() method,
 * thus is able to dispatch actions to the store.
 * We import setTextFilter to update the text filter everytime users type in the input
 *
 * @param props
 * @returns {*}
 * @constructor
 */
export class ExpensesListFilters extends React.Component {
	state = {
		calendarFocused: null
	}

	onDatesChange = ({ startDate, endDate }) => {
		this.props.setStartDate(startDate)
		this.props.setEndDate(endDate)
	}

	onFocusChange = (calendarFocused) => {
		this.setState(() => ({ calendarFocused }))
	}

	onTextChange = (e) => {
		this.props.setTextFilter(e.target.value)
	}

	onSortChange = (e) => {
		if (e.target.value === 'amount') {
			this.props.sortByAmount()
		} else if (e.target.value === 'date') {
			this.props.sortByDate()
		}
	}

	render () {
		return (
			<div className="input-group">
				<div className="input-group__item">
					<input
						type="text"
						defaultValue={this.props.filters.text}
						onChange={this.onTextChange}
						placeholder="Search expenses"
						className="text-input"
					/>
				</div>

				<div className="input-group__item">
					<select
						value={this.props.filters.sortBy}
						onChange={this.onSortChange}
						className="select"
					>
						<option value="date">Date</option>
						<option value="amount">Amount</option>
					</select>
				</div>

				<div className="input-group__item">
					<DateRangePicker
						startDate={this.props.filters.startDate}
						endDate={this.props.filters.endDate}
						startDateId="startDate"
						endDateId="endDate"
						onDatesChange={this.onDatesChange}
						focusedInput={this.state.calendarFocused}
						onFocusChange={this.onFocusChange}
						numberOfMonths={1}
						isOutsideRange={() => false}
						showClearDates={true}
					/>
				</div>
			</div>
		)
	}
}

const mapStateToProps = (state) => ({ filters: state.filters })

const mapDispatchToProps = (dispatch) => ({
	setTextFilter: (text) => dispatch(setTextFilter(text)),
	sortByDate: () => dispatch(sortByDate()),
	sortByAmount: () => dispatch(sortByAmount()),
	setStartDate: (startDate) => dispatch(setStartDate(startDate)),
	setEndDate: (endDate) => dispatch(setEndDate(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpensesListFilters)