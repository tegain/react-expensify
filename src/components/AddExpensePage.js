import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { addExpense } from '../actions/expensesActions'

export class AddExpensePage extends React.Component {
	onSubmit = (expense) => {
		this.props.addExpense(expense) // Dispatch addExpense to the store
		this.props.history.push('/') // Redirects to the dashboard page
	}

	render () {
		return (
			<div>
				<h1>Add Expense</h1>
				<ExpenseForm
					onSubmit={this.onSubmit}
				/>
			</div>
		)
	}
}

// Use mapDispatchToProps to make the dispatch action independant from the comonent itself
// It makes the component easier to test, because there is only one linked method to test
const mapDispatchToProps = (dispatch) => ({
	addExpense: (expense) =>Â dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage)