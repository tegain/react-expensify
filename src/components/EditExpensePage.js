import React from 'react'
import { connect } from 'react-redux'
import {startEditExpense, startRemoveExpense} from '../actions/expensesActions'
import ExpenseForm from './ExpenseForm'

export class EditExpensePage extends React.Component {
	onSubmit = (expense) => {
		this.props.startEditExpense(this.props.expense.id, expense)
		this.props.history.push('/')
	}

	onRemove = () => {
		this.props.startRemoveExpense({ id: this.props.expense.id })
		this.props.history.push('/')
	}

	render () {
		return (
			<div>
				<div className="content-header">
					<div className="content-header__container container">
						<h2 className="content-header__heading">Edit expense</h2>
					</div>
				</div>

				<div className="content-body">
					<div className="content-body__container container">
						<ExpenseForm
							expense={this.props.expense}
							onSubmit={this.onSubmit}
						/>
						<button
							onClick={this.onRemove}
							className="button button--ghost"
						>
							Remove expense
						</button>
					</div>
				</div>
			</div>
		)
	}
}

const mapStateToProps = (state, props) => ({
	expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch, props) => ({
	startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
	startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)