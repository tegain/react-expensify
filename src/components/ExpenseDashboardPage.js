import React from 'react'
import ExpensesSummary from './ExpensesSummary'
import ExpensesList from './ExpensesList'
import ExpensesListFilters from './ExpensesListFilters'

const ExpenseDashboardPage = () => (
	<div>
		<ExpensesSummary />
		<ExpensesListFilters />
		<ExpensesList />
	</div>
)

export default ExpenseDashboardPage