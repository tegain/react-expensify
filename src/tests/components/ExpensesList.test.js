import React from 'react'
import { shallow } from 'enzyme'
import { ExpensesList } from '../../components/ExpensesList'
import expenses from '../fixtures/expenses'

test('should render <ExpensesList /> with expenses', () => {
	const wrapper = shallow(<ExpensesList expenses={expenses} />)
	expect(wrapper).toMatchSnapshot()
})

test('should render ExpensesList with empty message', () => {
	const wrapper = shallow(<ExpensesList expenses={[]} />)
	expect(wrapper).toMatchSnapshot()
})