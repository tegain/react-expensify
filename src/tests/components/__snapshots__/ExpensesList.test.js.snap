// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render <ExpensesList /> with expenses 1`] = `
<div>
  <ExpensesListItem
    amount={195}
    createdAt={0}
    description="Rent"
    id={1}
    key="1"
    note=""
  />
  <ExpensesListItem
    amount={1050}
    createdAt={-345600000}
    description="Gas bill"
    id={2}
    key="2"
    note="That's pretty expensive!"
  />
  <ExpensesListItem
    amount={825}
    createdAt={345600000}
    description="Internet bill"
    id={3}
    key="3"
    note=""
  />
</div>
`;

exports[`should render ExpensesList with empty message 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;
